src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup.hooks.tsx

```diff
@@ -0,0 +1,146 @@
+import { NUM_WORKERS } from './DataExportWorker/DataExportWorker.const';
+
+export const useMultipleDecryptWorkers = ({
+  handleExportPopupClose,
+  appletId,
+  encryption,
+  filters,
+}: MultipleDecryptWorkersProps) => {
+  const [dataIsExporting, setDataIsExporting] = useState(false);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [exportDataQueue, setExportDataQueue] = useState<ExportDataResult[] | null>(null);
+
+  const dataExportingApiRef = useRef(false);
+  const workersRef = useRef<IdleWorker[]>([]);
+  const processedPagesRef = useRef<Set<number>>(new Set());
+  const finishedPagesRef = useRef<Set<number>>(new Set());
+  const limitRef = useRef(1);
+  const taskQueueRef = useRef<{ pageData: ExportDataResult; page: number }[]>([]);
+  const shouldLogDataInDebugMode =
+    !isProduction && sessionStorage.getItem(SessionStorageKeys.DebugMode) === 'true';
+
+  const { getAppletPrivateKey } = useEncryptionStorage();
+
+  const encryptionInfoFromServer = getParsedEncryptionFromServer(encryption);
+  const privateKey = getAppletPrivateKey(appletId) as number[];
+
+  const distributeTasks = useCallback(() => {
+    workersRef.current.forEach((worker) => {
+      if (worker.isIdle && taskQueueRef.current.length > 0) {
+        const currentTask = taskQueueRef.current.shift();
+        if (!currentTask) return;
+
+        const { pageData, page } = currentTask;
+        worker.isIdle = false;
+        worker.postMessage({
+          encryptedData: pageData,
+          encryptionInfoFromServer,
+          page,
+          privateKey,
+          hasSuffix: limitRef.current > 1,
+          filters,
+          shouldLogDataInDebugMode,
+          ItemResponseType,
+        });
+
+        setCurrentPage((prevState) => prevState + 1);
+      }
+    });
+  }, [encryptionInfoFromServer, privateKey, filters, shouldLogDataInDebugMode]);
+
+  const processData = useCallback(
+    (pageData: ExportDataResult, page: number) => {
+      if (workersRef.current.length === 0 || processedPagesRef.current.has(page)) return;
+
+      taskQueueRef.current.push({ pageData, page });
+      distributeTasks();
+    },
+    [distributeTasks],
+  );
+
+  const handleAllTasksCompleted = () => {
+    // check if all pages are processed and all workers are idle
+    const allPagesProcessed =
+      processedPagesRef.current.size === limitRef.current &&
+      finishedPagesRef.current.size === limitRef.current;
+    const allWorkersIdle = workersRef.current.every((worker) => worker.isIdle);
+
+    if (!dataExportingApiRef.current && allPagesProcessed && allWorkersIdle) {
+      setDataIsExporting(false);
+      handleExportPopupClose();
+      Mixpanel.track('Export Data Successful', {
+        'Applet ID': appletId,
+      });
+    }
+  };
+
+  useEffect(() => {
+    if (!exportDataQueue) return;
+
+    const processingBatch = exportDataQueue[currentPage - 1];
+
+    if (processingBatch) {
+      processData(processingBatch, currentPage);
```

Comment:
`useEffect` should only be used to sync components with an external system:
https://react.dev/reference/react/useEffect

In this case, there is no synchronization with an external API. Hence, it should work the other way. For instance, a user event should invoke the `processData` (e.g. click).

Reply:
There is no user event to invoke `processData` because the goal is to fetch all the pages of export data sequentially. This happens automatically as each page is received, without requiring user interaction. By using `useEffect`, we ensure that the data fetching and subsequent decryption/processing happen continuously and in parallel. This approach allows us to handle the data processing efficiently, as each page triggers the next step in the sequence without waiting for a user action.